### PR TITLE
tmem: Add tmemCloseHandle(), tmemWaitForPermission(), use those functions in nv.c to fix a race condition

### PR DESCRIPTION
**Pull request by masagrator**
When TransferMemory has different permission than RW, we need to check if `src_addr` permission was restored to original state as we currently don't have any assurance this happened when `__libnx_free` is called.

This is fixing an issue with Tesla overlays crashing randomly (mainly in docked mode) related to this bug.

Thanks to `~SciresM` for providing an idea to solve this and `~HE` from ReSwitched for help with shaping out code.

### CODE REVIEW

**Comment by fincs** [comment]
Since this is a bug caused by a race condition in nvservices, it seems a bad idea to special case the workaround in tmemClose. An alternative solution would be to introduce a new `tmemWaitForPermission` function that would wait for a given tmem object to match the specified permission mask/value (as parameters); and use that in nv.c just before calling tmemClose.

**Comment by masagrator** [comment]
This cannot be done before tmemUnmap is used, otherwise it will stuck in infinite loop.

My proposition is introducing `tmemWaitForPermissionAndClose` that would do what currently my PR is doing, but as separate function.

**Comment by masagrator** [comment]
Then maybe it's a handle that must be closed first? If not, then I have no idea why it's not working.

My fork I was testing
https://github.com/switchbrew/libnx/compare/master...masagrator:libnx:patch-2

**Comment by yellows8** [comment]
Impl the `tmemWaitForPermissionAndClose` you mentioned.

`Then maybe it's a handle that must be closed first? If not, then I have no idea why it's not working.` <- The handle must be closed for that to do anything.

**Comment by yellows8** [comment]
Actually you can keep tmemWaitForPermission if tmemCloseHandle is added. tmemCloseHandle would have the handle-closing code from tmemClose, and you'd call this following the _nvCmdInitialize call.

**Comment by masagrator** [comment]
Checked this in overlays with 2 people and it works fine.
I guess this will need grammar fixes as I'm not great at this.

**Comment by yellows8** [comment]
Replace the duplicate code in tmemClose with a call to tmemCloseHandle.

File: nx/include/switch/kernel/tmem.h [hunk 1]
```diff
@@ -72,6 +72,23 @@ static inline void* tmemGetAddr(TransferMemory* t){
...
 }
 
+/**
+ * @brief Closes handle of a transfer memory object.
+ * @param t Transfer memory information structure.
+ * @return Result code.
+ */
+Result tmemCloseHandle(TransferMemory* t);
+
+
```
Comment (yellows8) [inline nx/include/switch/kernel/tmem.h:82]: Remove extra newline.

File: nx/include/switch/kernel/tmem.h [hunk 2]
```diff
@@ -72,6 +72,23 @@ static inline void* tmemGetAddr(TransferMemory* t){
...
+/**
+ * @brief Waits until source backing memory permissions match perm.
+ * @param t Transfer memory information structure.
+ * @param perm Permissions which the source backing memory is expected to have before return.
+ * @return Result code.
+ */
+Result tmemWaitForPermission(TransferMemory* t, Permission perm);
+
+
```
Comment (yellows8) [inline nx/include/switch/kernel/tmem.h:91]: Remove extra newline.

File: nx/source/services/nv.c [hunk 3]
```diff
@@ -54,6 +54,9 @@ Result _nvInitialize(void) {
 
         if (R_SUCCEEDED(rc))
             rc = _nvCmdInitialize(CUR_PROCESS_HANDLE, g_nvTransfermem.handle, tmem_size);
+        
+        if (R_SUCCEEDED(rc))
```
Comment (yellows8) [inline nx/source/services/nv.c:58]: This has to be unconditional. Something like: `Result rc2 = tmemCloseHandle(...);` `if (R_SUCCEEDED(rc)) rc = rc2;`

**Review by yellows8 — CHANGES_REQUESTED** [review]

**Review by yellows8 — APPROVED** [review]

**Review by fincs — APPROVED** [review]

**Comment by masagrator** [comment]
If such hang happens, you should attach GDB to it and check if it's not stuck in loop. For example in tmemWaitForPermission().
But if this is true, then it would mean that without this function your app would crash with 0x2A8 error code.

**Comment by masagrator** [comment]
I'm not sure if only RW or RW and R since this is something platform specific. 

You can try to write Perm_R instead of Perm_Rw and see what will happen, it should catch any Read permission.

**Pull request merged**